### ISSUE TITLE
FR: yaml value suggestions

### ISSUE THREAD

**Issue by chrisgrieser** [tadashi-aikawa/obsidian-various-complements-plugin#74]
similar to #48, but only more general: enable yaml-key-specific complements. For example, if I regularly use 3-4 `cssclass` yaml values; so it would be nice if the plugin could simply show me a suggester with those 4. 

Having autocomplete in the yaml frontmatter is quite valuable, since it ensures consistent assignment of yaml values.

**Comment by chrisgrieser** [comment]
like here is also a nice idea, too: https://github.com/SkepticMystic/codeblock-completer

**Comment by tadashi-aikawa** [comment]
Hi, @chrisgrieser . Sorry for the late reply.

I think it is a little ways away, but I want to try it after #73, #16, #48, etc. :)

**Comment by chrisgrieser** [comment]
no worries, it's just something for far down the line. :)

**Comment by tadashi-aikawa** [comment]
Hi, @chrisgrieser 
I have released [v5.6.0.beta3](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin-beta/releases/tag/5.6.0.beta3) 🚀 

![](https://user-images.githubusercontent.com/9500018/153822780-d7dfe475-1818-4959-afdc-7491a3613a00.gif)

It is a prototype version, so I think there are some bugs and points to improve. I don't often use YAML front matter (except for `aliases`), so your feedback would be helpful for me.

Kind regards :)

**Comment by chrisgrieser** [comment]
thanks 🥰 

So I gave it a spin and I see some things that could be improved:
1) the file name should be made optional again, in many cases, I don't care where the suggestion comes from.
2) it should only suggest individual values, not a list of values
3) after completion, it should add `, ` so you can keep on adding more values.
4) it should suggest items immediately, not just after n letters (n being the number of letters you have set in the settings.) Reason being that yaml values are often highly standardized, so as opposed to normal text, autosuggestion disturb less and should start earlier.
5) right now, only the *values* are autosuggested. However, it could also make sense to suggest keys itself when you are on a new line in the yaml? So I don't have to type out thinks like `cssclass` all the time.

<img width="728" alt="Screenshot 2022-02-14 10 28 13" src="https://user-images.githubusercontent.com/73286100/153836925-2be9cb3b-98ba-4bf4-a785-bd42b149513b.png">

**Comment by tadashi-aikawa** [comment]
@chrisgrieser 
Thanks for your feedback!

> 1. the file name should be made optional again, in many cases, I don't care where the suggestion comes from.

Exactly. In the case of Front matter, I think it is not necessary to show a description.

> 2. it should only suggest individual values, not a list of values

I couldn't reproduce it. Don't you write a list not supported for YAML format? (ex: `key: value1, value2, value3`)

> 3. after completion, it should add , so you can keep on adding more values.

I'll consider it an option because it doesn't need a comma to write another YAML format of an array.

ex.
```
tags:
  - one
  - two
```

> 4. it should suggest items immediately, not just after n letters (n being the number of letters you have set in the settings.) Reason being that yaml values are often highly standardized, so as opposed to normal text, autosuggestion disturb less and should start earlier.

I agree perfectly 👍 

> 5. right now, only the values are autosuggested. However, it could also make sense to suggest keys itself when you are on a new line in the yaml? So I don't have to type out thinks like cssclass all the time.

That is so useful 😄 I want to implement it after v5.6.0 :)

**Comment by tadashi-aikawa** [comment]
Released [v5.6.0.beta4](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin-beta/releases/tag/5.6.0.beta4) 🚀 

Fix 1 and 4 🎉

**Comment by chrisgrieser** [comment]
>> it should only suggest individual values, not a list of values
> I couldn't reproduce it. Don't you write a list not supported for YAML format? (ex: key: value1, value2, value3)

see the screenshot above – `key: value1, value2, value3` is what I do. And it suggests to my `value1, value2, value3` instead of `value1`

**Comment by chrisgrieser** [comment]
yeah, but Obsidian is more lenient and accepts those kind of yaml, though: 
<img width="549" alt="Screenshot 2022-02-14 12 52 52" src="https://user-images.githubusercontent.com/73286100/153859616-f0ded6f6-a64d-4316-a72c-d54576b722cf.png">

**Comment by tadashi-aikawa** [comment]
Thanks!

Considering other than Obsidian, it shouldn't, but Various Complements is a plugin for Obsidian, so I'll support it :+1:

**Comment by tadashi-aikawa** [comment]
- `tags` and `tag` are both valid
- `aliases` and `alias` are both valid

Ah.. That is why there is a particular API only for tags and aliases. I understood.

**Comment by tadashi-aikawa** [comment]
@chrisgrieser 
Released [v5.6.0.beta5](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin-beta/releases/tag/5.6.0.beta5) 🚀 

It would support all cases for Tag and Alias 👍

```
tag: aaa, bbb
tag: [aaa, bbb]
tag:
  - aaa
  - bbb

tags: aaa, bbb
tags: [aaa, bbb]
tags:
  - aaa
  - bbb

alias: aaa, bbb
alias: [aaa, bbb]
alias:
  - aaa
  - bbb

aliases: aaa, bbb
aliases: [aaa, bbb]
aliases:
  - aaa
  - bbb
```

**Comment by chrisgrieser** [comment]
thanks 🤩 

btw, it's not only tags and aliases, I just tested it, this "loose" format is also supported for any yaml key you choose, so it might make sense to implement it for everything?

**Comment by tadashi-aikawa** [comment]
Yes, I have found that `parseFrontMatterStringArray` seems to make us possible 👍  just a moment please.

**Comment by tadashi-aikawa** [comment]
Released [v5.6.0.beta6](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin-beta/releases/tag/5.6.0.beta6) 🚀

By the way, I'm worried about the performance a little.  It is fine in my environment (indexing time is less than 10ms), but isn't it fine with much front matter in the vault?

**Comment by chrisgrieser** [comment]
Well, I tried it out a bit, but I haven't noticed big problems yet regarding performance yet.

it works better now, but the non-tag-yaml keys suggest non-yaml values 
<img width="504" alt="Screenshot 2022-02-14 14 18 00" src="https://user-images.githubusercontent.com/73286100/153871882-7da99f86-e8c7-44b4-9627-b8f9cb8d9649.png">

---

this leaves pretty much [only 3 & 5](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/issues/74#issuecomment-1038860821), otherwise it's perfect already :)

---

also, another thing to think about is how would interact with other plugins that also autosuggest tags, [like this one](https://github.com/jmilldotdev/obsidian-frontmatter-tag-suggest).

**Comment by tadashi-aikawa** [comment]
Released [v5.6.0.beta7](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin-beta/releases/tag/5.6.0.beta7) 🚀

**Comment by chrisgrieser** [comment]
works, well, haven't noticed anything for now :)

**Comment by tadashi-aikawa** [comment]
>after completion, it should add , so you can keep on adding more values.

Added a `Insert comma after completion` option 👍 

![2022-02-19](https://user-images.githubusercontent.com/9500018/154787499-c0d96709-b990-4e8a-9ccf-96629bf465b3.gif)

**Comment by tadashi-aikawa** [comment]
Before the v5.6.0 official release, I have implemented the last improvement, which suggests property values even if we don't type anything yet.

It is helpful to see selections without filtering the first character. Primarily, In the case that there are not too many items.

![2022-02-19](https://user-images.githubusercontent.com/9500018/154791907-4028fb24-517d-4c4f-8825-ab6027a94349.gif)

We don't need to worry about the performance because it merely suggest s items equals to `Max number of suggestions`.

**Comment by tadashi-aikawa** [comment]
I have released [v5.6.0](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/releases/tag/5.6.0) 🚀

Many thanks!
